File: src/tokens.jl

typealias String ASCIIString

global words = Dict()
global tags = Dict()
global labels = Dict()
global FORMAT = nothing

immutable Word
    wordstr  ::String
    word     ::Int
    tagstr   ::String
    tag      ::Int
    head     ::Int
    labelstr ::String
    label    ::Int
    feature
    function Word(word, tag, head, label, feature=nothing)
        word_id  = get!(words, word, length(words))
        tag_id   = get!(tags, tag, length(tags))
        label_id = get!(labels, label, length(labels))
        new(word, word_id, tag, tag_id, head, label, label_id, feature)
    end
end

typealias Sent Vector{Word}

rootword = Word("root", "", 0, "")

import Base.(==)
(==)(w1::Word, w2::Word) = w1.word == w2.word

heads(s::Vector{Word}) = map(s->s.head, s)

function Word(line::String)
    line     = line |> chomp |> split
    word     = line[FORMAT["word"]]
    tag      = line[FORMAT["tag"]]
    head     = parse(Int, line[FORMAT["head"]])
    label    = line[FORMAT["label"]]
    Word(word, tag, head, label)
end

Base.print(io::IO, w::Word) =
    print(io, join([w.wordstr, "_", w.tagstr,
      w.tagstr, "_", w.head, w.labelstr, "_", "_"], "\t"))

function Base.print(io::IO, sent::Sent)
    for i = 1:length(sent)
        println(io, i, "\t", sent[i])
    end
end

function readconll(filename::String)
    res = Sent[[]]
    for line in open(readlines, filename)
        if length(line) < 3
            push!(res, []); continue
        end
        push!(res[end], Word(line))
    end
    res
end

macro ConllFormat(args...)
    i = 0
    res = Dict{String,Int}()
    while length(args) > i
        i += 1
        args[i] == :(:-) && continue
        res[string(args[i])] = i
    end
    quote global FORMAT = $(esc(res)) end
end

function dump(filename::String, sents::Vector{Sent})
    res = open(filename, "w")
    for sent in sents
        println(res, sent)
    end
    close(res)
end

File: cpp/tokens.cpp
#include "tokens.h"

using namespace std;

#define print(var) std::cout<<#var"= "<<var<<std::endl
unordered_map<string,int> words;
unordered_map<string,int> tags;
unordered_map<string,int> labels;
unordered_map<string,int> format;

string root = "root";

int getOrAdd(unordered_map<string,int>& dict, string& key) {
    int id = dict[key];
    if (id == 0) {
        id = dict.size() + 1;
        dict[key] = id;
    }
    return id;
}

Word::Word(string& _word, string& _pos, string& _head, string& _label) {
    init(_word, _pos, stoi(_head), _label);
}
Word::Word() { init(root, root, 0, root); }

Word::Word(vector<string>& line) {
    init(line[format["word"]],
         line[format["pos"]],
         stoi(line[format["head"]]),
         line[format["label"]]);
}

void Word::init(string& _word, string& _pos, int _head, string& _label) {
    word  = _word;
    pos   = _pos;
    head  = _head;
    label = _label;
    w = getOrAdd(words, word);
    p = getOrAdd(tags, pos);
    l = getOrAdd(labels, label);
}

string Word::toConll() {
    ostringstream os;
    os << word  << "\t"
       << "_"   << "\t"
       << pos   << "\t"
       << pos   << "\t"
       << "_"   << "\t"
       << head  << "\t"
       << label << "\t"
       << "_"   << "\t"
       << "_";
    return os.str();
}

void toConll(ostream& os, vector<Word>& sent) {
    for (unsigned i = 1; i < sent.size(); i++)
        os << i << "\t" << sent[i].toConll() << endl;
    os << endl;
}
void toConll(vector<Word>& sent) { toConll(cout, sent); }

vector<vector<Word> > readConll(const string& file) {
    initFormat();
    ifstream conllFile(file);
    string line;
    Word root = Word();
    vector< vector<Word> > res;
    vector<Word> sent;
    sent.push_back(root);
    while (getline(conllFile, line)) {
        if (line.size() == 1) {
            res.push_back(sent);
            sent.clear();
            sent.push_back(root);
            continue;
        }
        vector<string> tmp = split(line, '\t');
        sent.push_back(Word(tmp));
    }
    return res;
}

void initFormat() {
    format["word"]  = 1;
    format["pos"]   = 3;
    format["head"]  = 6;
    format["label"] = 7;
}

int* getHeads(vector<Word>& sent) {
    int* res = new int[sent.size()];
    for (int i = 0; i < sent.size(); i++) {
        res[i] = sent[i].head;
    }
    return res;
}

// int main() {
//     const string file = "wsj_23.conll";
//     auto res = readConll(file);
//
//     for (unsigned i = 0; i < res.size(); i++) {
//         // toConll(cout, res[i]);
//         for (int j = 0; j < res[i].size(); j++) {
//             print(res[i][j].w);
//             print(res[i][j].p);
//             print(res[i][j].l);
//         }
//     }
//
//     // for (auto iter = words.begin(); iter != words.end(); iter++) {
//     //     cout << iter->first << " => " << iter->second << endl;
//     // }
// }
